File: apps/application/src/browser_harness/context.py
from __future__ import annotations

import asyncio
import json
import multiprocessing as mp
from dataclasses import dataclass
from types import MappingProxyType
from typing import Literal, Mapping

from .artifacts import StoredCandidateArtifacts, StoredUpload
from .models import HarnessServiceError, validate_sanitized_basename

MAX_SOURCE_CHARACTERS = 100_000
MAX_RESUME_SOURCE_CHARACTERS = 256 * 1024
_BASE_COMBINED_NARRATIVE_CHARACTERS = 250_000
MAX_COMBINED_NARRATIVE_CHARACTERS = (
    _BASE_COMBINED_NARRATIVE_CHARACTERS
    + MAX_RESUME_SOURCE_CHARACTERS
    - MAX_SOURCE_CHARACTERS
)

SourceCategory = Literal["profile", "context", "anecdote"]
EvidenceCategory = Literal["resume", "profile", "context", "anecdote"]


@dataclass(frozen=True, slots=True)
class AttributedSource:
    name: str
    category: SourceCategory
    text: str

    def __post_init__(self) -> None:
        validate_sanitized_basename(self.name)
        if len(self.text) > MAX_SOURCE_CHARACTERS:
            raise ValueError("candidate source exceeds the character limit")


@dataclass(frozen=True, slots=True)
class CandidateContext:
    direct_fields: Mapping[str, str]
    resume_text: str
    profile_narrative: AttributedSource
    context_sources: tuple[AttributedSource, ...]
    anecdotes: tuple[AttributedSource, ...]


@dataclass(frozen=True, slots=True)
class AttributedEvidence:
    category: EvidenceCategory
    name: str
    text: str

    def __post_init__(self) -> None:
        if self.category not in {"resume", "profile", "context", "anecdote"}:
            raise ValueError("candidate evidence has an invalid category")
        validate_sanitized_basename(self.name)
        maximum_characters = (
            MAX_RESUME_SOURCE_CHARACTERS
            if self.category == "resume"
            else MAX_SOURCE_CHARACTERS
        )
        if len(self.text) > maximum_characters:
            raise ValueError("candidate evidence exceeds the character limit")

    def as_task_value(self) -> dict[str, str]:
        return {
            "category": self.category,
            "name": self.name,
            "text": self.text,
        }


def _invalid_context() -> HarnessServiceError:
    return HarnessServiceError(422, "invalid_request", "Candidate context is invalid")


def _read_utf8(
    upload: StoredUpload,
    *,
    maximum_characters: int = MAX_SOURCE_CHARACTERS,
) -> str:
    try:
        text = upload.path.read_text(encoding="utf-8")
    except (OSError, UnicodeError):
        raise _invalid_context() from None
    if len(text) > maximum_characters:
        raise _invalid_context()
    return text


def load_candidate_context(artifacts: StoredCandidateArtifacts) -> CandidateContext:
    """Extract bounded, attributed evidence without placing direct fields in prompt text."""

    try:
        resume_text = _read_utf8(
            artifacts.resume_source,
            maximum_characters=MAX_RESUME_SOURCE_CHARACTERS,
        )
        profile = AttributedSource(
            name=artifacts.personal_upload.display_name,
            category="profile",
            text=artifacts.personal.narrative,
        )
        contexts = tuple(
            AttributedSource(name=upload.display_name, category="context", text=_read_utf8(upload))
            for upload in artifacts.contexts
        )
        anecdotes = tuple(
            AttributedSource(name=upload.display_name, category="anecdote", text=_read_utf8(upload))
            for upload in artifacts.anecdotes
        )
        combined_length = (
            len(resume_text)
            + len(profile.text)
            + sum(len(source.text) for source in contexts)
            + sum(len(source.text) for source in anecdotes)
        )
        combined_limit = _BASE_COMBINED_NARRATIVE_CHARACTERS + max(
            0,
            len(resume_text) - MAX_SOURCE_CHARACTERS,
        )
        if combined_length > combined_limit:
            raise _invalid_context()
        direct_fields = MappingProxyType(dict(artifacts.personal.direct_fields))
        return CandidateContext(
            direct_fields=direct_fields,
            resume_text=resume_text,
            profile_narrative=profile,
            context_sources=contexts,
            anecdotes=anecdotes,
        )
    except HarnessServiceError:
        raise
    except (OSError, UnicodeError, ValueError):
        raise _invalid_context() from None


def candidate_evidence_records(
    candidate: CandidateContext,
    resume_name: str,
) -> tuple[AttributedEvidence, ...]:
    records = [
        AttributedEvidence(
            category="resume",
            name=resume_name,
            text=candidate.resume_text,
        )
    ]
    if candidate.profile_narrative.text:
        records.append(
            AttributedEvidence(
                category=candidate.profile_narrative.category,
                name=candidate.profile_narrative.name,
                text=candidate.profile_narrative.text,
            )
        )
    records.extend(
        AttributedEvidence(
            category=source.category,
            name=source.name,
            text=source.text,
        )
        for source in candidate.context_sources
    )
    records.extend(
        AttributedEvidence(
            category=source.category,
            name=source.name,
            text=source.text,
        )
        for source in candidate.anecdotes
    )
    return tuple(records)


def render_candidate_evidence(candidate: CandidateContext, resume_name: str) -> str:
    """Render the retained JSONL debug view of attributed evidence."""

    sections = (
        json.dumps(
            record.as_task_value(),
            ensure_ascii=False,
            separators=(",", ":"),
        )
        for record in candidate_evidence_records(candidate, resume_name)
    )
    return "Candidate evidence sources (one JSON object per line):\n" + "\n".join(
        sections
    )


def _serialize_candidate_context(candidate: CandidateContext) -> dict[str, object]:
    def source(value: AttributedSource) -> dict[str, str]:
        return {
            "name": value.name,
            "category": value.category,
            "text": value.text,
        }

    return {
        "ok": True,
        "direct_fields": list(candidate.direct_fields.items()),
        "resume_text": candidate.resume_text,
        "profile_narrative": source(candidate.profile_narrative),
        "context_sources": [source(value) for value in candidate.context_sources],
        "anecdotes": [source(value) for value in candidate.anecdotes],
    }


def _candidate_context_worker(
    artifacts: StoredCandidateArtifacts,
    sender: Connection,
) -> None:
    try:
        try:
            payload = _serialize_candidate_context(load_candidate_context(artifacts))
        except Exception:
            payload = {"ok": False}
        sender.send(payload)
    finally:
        sender.close()


def _source_from_payload(value: object) -> AttributedSource:
    if not isinstance(value, dict):
        raise _invalid_context()
    try:
        return AttributedSource(
            name=value["name"],
            category=value["category"],
            text=value["text"],
        )
    except (KeyError, TypeError, ValueError):
        raise _invalid_context() from None


def _candidate_from_payload(payload: object) -> CandidateContext:
    if not isinstance(payload, dict) or payload.get("ok") is not True:
        raise _invalid_context()
    try:
        direct_items = payload["direct_fields"]
        if not isinstance(direct_items, list):
            raise TypeError
        direct_fields = dict(direct_items)
        if any(
            not isinstance(name, str) or not isinstance(value, str)
            for name, value in direct_fields.items()
        ):
            raise TypeError
        resume_text = payload["resume_text"]
        context_values = payload["context_sources"]
        anecdote_values = payload["anecdotes"]
        if (
            not isinstance(resume_text, str)
            or not isinstance(context_values, list)
            or not isinstance(anecdote_values, list)
        ):
            raise TypeError
        return CandidateContext(
            direct_fields=MappingProxyType(direct_fields),
            resume_text=resume_text,
            profile_narrative=_source_from_payload(payload["profile_narrative"]),
            context_sources=tuple(
                _source_from_payload(value) for value in context_values
            ),
            anecdotes=tuple(
                _source_from_payload(value) for value in anecdote_values
            ),
        )
    except (HarnessServiceError, KeyError, TypeError, ValueError):
        raise _invalid_context() from None


class CandidateContextProcess:
    """Extract candidate evidence in a killable subprocess."""

    def __init__(self, artifacts: StoredCandidateArtifacts) -> None:
        context = mp.get_context("spawn")
        receiver, sender = context.Pipe(duplex=False)
        process = context.Process(
            target=_candidate_context_worker,
            args=(artifacts, sender),
            name="jobhunter-candidate-context",
            daemon=True,
        )
        try:
            process.start()
        except Exception:
            receiver.close()
            sender.close()
            raise _invalid_context() from None
        sender.close()
        self._receiver = receiver
        self._process = process
        self._receive_task: asyncio.Task[object] | None = None
        self._closed = False

    async def result(self) -> CandidateContext:
        if self._closed:
            raise _invalid_context()
        if self._receive_task is None:
            self._receive_task = asyncio.create_task(
                asyncio.to_thread(self._receiver.recv)
            )
        try:
            payload = await asyncio.shield(self._receive_task)
        except (EOFError, OSError):
            raise _invalid_context() from None
        await self._join_or_terminate()
        return _candidate_from_payload(payload)

    async def terminate(self) -> None:
        if self._closed:
            return
        if self._process.is_alive():
            self._process.terminate()
        self._receiver.close()
        if self._receive_task is not None:
            try:
                await asyncio.shield(self._receive_task)
            except (EOFError, OSError):
                pass
        await self._join_or_terminate()

    async def _join_or_terminate(self) -> None:
        if self._closed:
            return
        await asyncio.to_thread(self._process.join, 5)
        if self._process.is_alive():
            self._process.kill()
            await asyncio.to_thread(self._process.join, 5)
        self._receiver.close()
        self._process.close()
        self._closed = True

File: apps/harness/src/application/context.ts
import { readFile } from "node:fs/promises";

import { HarnessServiceError, type StoredCandidateArtifacts, type StoredUpload } from "../host/artifacts.ts";

export const MAX_SOURCE_CHARACTERS = 500_000;
export const MAX_RESUME_SOURCE_CHARACTERS = 1_310_720;
const BASE_COMBINED_NARRATIVE_CHARACTERS = 1_250_000;
export const MAX_COMBINED_NARRATIVE_CHARACTERS = 2_060_720;

export type SourceCategory = "profile" | "context" | "anecdote";
export type EvidenceCategory = "resume" | SourceCategory;

export interface AttributedSource {
  readonly name: string;
  readonly category: SourceCategory;
  readonly text: string;
}
export interface CandidateContext {
  readonly directFields: Readonly<Record<string, string>>;
  readonly resumeText: string;
  readonly profileNarrative: AttributedSource;
  readonly contextSources: readonly AttributedSource[];
  readonly anecdotes: readonly AttributedSource[];
}
export interface AttributedEvidence {
  readonly category: EvidenceCategory;
  readonly name: string;
  readonly text: string;
}

function invalidContext(): HarnessServiceError {
  return new HarnessServiceError(422, "invalid_request", "Candidate context is invalid");
}

function characterCount(value: string): number {
  let count = 0;
  for (const _character of value) count += 1;
  return count;
}

function validateSanitizedBasename(value: string): string {
  if (typeof value !== "string") throw new TypeError("filename must be a string");
  const length = characterCount(value);
  if (length < 1 || length > 255) throw new Error("filename must contain 1 to 255 characters");
  if (value === "." || value === ".." || value.includes("/") || value.includes("\\")) throw new Error("filename must be a basename");
  for (const character of value) if ((character.codePointAt(0) ?? 0) < 32) throw new Error("filename contains unsafe characters");
  return value;
}

function attributedSource(name: string, category: SourceCategory, text: string): AttributedSource {
  validateSanitizedBasename(name);
  if (characterCount(text) > MAX_SOURCE_CHARACTERS) throw new Error("candidate source exceeds the character limit");
  return Object.freeze({ name, category, text });
}

async function readUtf8(upload: StoredUpload, maximumCharacters = MAX_SOURCE_CHARACTERS): Promise<string> {
  const bytes = await readFile(upload.path);
  const text = new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  if (characterCount(text) > maximumCharacters) throw new Error("candidate source exceeds the character limit");
  return text;
}

export async function loadCandidateContext(artifacts: StoredCandidateArtifacts): Promise<CandidateContext> {
  try {
    const resumeText = await readUtf8(artifacts.resumeSource, MAX_RESUME_SOURCE_CHARACTERS);
    const profileNarrative = attributedSource(artifacts.personalUpload.displayName, "profile", artifacts.personal.narrative);
    const contextSources = await Promise.all(artifacts.contexts.map(async (upload) => attributedSource(upload.displayName, "context", await readUtf8(upload))));
    const anecdotes = await Promise.all(artifacts.anecdotes.map(async (upload) => attributedSource(upload.displayName, "anecdote", await readUtf8(upload))));
    const resumeLength = characterCount(resumeText);
    const combinedLength = resumeLength + characterCount(profileNarrative.text)
      + contextSources.reduce((total, source) => total + characterCount(source.text), 0)
      + anecdotes.reduce((total, source) => total + characterCount(source.text), 0);
    const combinedLimit = BASE_COMBINED_NARRATIVE_CHARACTERS + Math.max(0, resumeLength - MAX_SOURCE_CHARACTERS);
    if (combinedLength > combinedLimit) throw invalidContext();
    const directFields: Record<string, string> = Object.create(null) as Record<string, string>;
    for (const [name, value] of artifacts.personal.directFields) directFields[name] = value;
    return Object.freeze({
      directFields: Object.freeze(directFields),
      resumeText,
      profileNarrative,
      contextSources: Object.freeze(contextSources),
      anecdotes: Object.freeze(anecdotes),
    });
  } catch (error) {
    if (error instanceof HarnessServiceError) throw error;
    throw invalidContext();
  }
}

export function candidateEvidenceRecords(candidate: CandidateContext, resumeName: string): readonly AttributedEvidence[] {
  const records: AttributedEvidence[] = [Object.freeze({ category: "resume", name: validateSanitizedBasename(resumeName), text: candidate.resumeText })];
  if (candidate.profileNarrative.text.length > 0) records.push(Object.freeze({
    category: candidate.profileNarrative.category,
    name: validateSanitizedBasename(candidate.profileNarrative.name),
    text: candidate.profileNarrative.text,
  }));
  for (const source of [...candidate.contextSources, ...candidate.anecdotes]) {
    records.push(Object.freeze({ category: source.category, name: validateSanitizedBasename(source.name), text: source.text }));
  }
  for (const record of records) {
    if (!(["resume", "profile", "context", "anecdote"] as const).includes(record.category)) throw new Error("candidate evidence has an invalid category");
    const limit = record.category === "resume" ? MAX_RESUME_SOURCE_CHARACTERS : MAX_SOURCE_CHARACTERS;
    if (characterCount(record.text) > limit) throw new Error("candidate evidence exceeds the character limit");
  }
  return Object.freeze(records);
}

export function renderCandidateEvidence(candidate: CandidateContext, resumeName: string): string {
  const records = candidateEvidenceRecords(candidate, resumeName);
  return "Candidate evidence sources (one JSON object per line):\n" + records.map((record) => JSON.stringify({
    category: record.category,
    name: record.name,
    text: record.text,
  })).join("\n");
}
export async function extractPdfText(path: string, signal?: AbortSignal): Promise<string> {
  signal?.throwIfAborted();
  const subprocess = Bun.spawn(["pdftotext", path, "-"], {
    stdout: "pipe",
    stderr: "ignore",
  });
  const abort = (): void => subprocess.kill();
  signal?.addEventListener("abort", abort, { once: true });
  try {
    const [exitCode, stdout] = await Promise.all([
      subprocess.exited,
      new Response(subprocess.stdout).text(),
    ]);
    signal?.throwIfAborted();
    if (exitCode !== 0) throw invalidContext();
    const text = stdout.replaceAll("\f", "").trimEnd();
    if (characterCount(text) > MAX_RESUME_SOURCE_CHARACTERS) throw invalidContext();
    return text;
  } catch (error) {
    if (error instanceof DOMException && error.name === "AbortError") throw error;
    throw invalidContext();
  } finally {
    signal?.removeEventListener("abort", abort);
  }
}
interface CandidateContextPayload {
  readonly ok: boolean;
  readonly candidate?: CandidateContext;
}

async function runContextWorker(): Promise<void> {
  try {
    const artifacts = JSON.parse(await Bun.stdin.text()) as StoredCandidateArtifacts;
    const candidate = await loadCandidateContext(artifacts);
    process.stdout.write(JSON.stringify({ ok: true, candidate } satisfies CandidateContextPayload));
  } catch {
    process.stdout.write(JSON.stringify({ ok: false } satisfies CandidateContextPayload));
  }
}

export class CandidateContextProcess {
  readonly #subprocess: Bun.Subprocess<"pipe", "pipe", "ignore">;
  #resultPromise: Promise<CandidateContext> | undefined;
  #closed = false;

  constructor(artifacts: StoredCandidateArtifacts) {
    try {
      this.#subprocess = Bun.spawn([process.execPath, import.meta.path, "--candidate-context-worker"], {
        stdin: "pipe",
        stdout: "pipe",
        stderr: "ignore",
      });
      const stdin = this.#subprocess.stdin;
      if (stdin === undefined) throw new Error("failed to open context worker input");
      stdin.write(JSON.stringify(artifacts));
      stdin.end();
    } catch {
      throw invalidContext();
    }
  }

  async result(): Promise<CandidateContext> {
    if (this.#closed) throw invalidContext();
    this.#resultPromise ??= this.#receive();
    try {
      const candidate = await this.#resultPromise;
      this.#closed = true;
      return candidate;
    } catch (error) {
      this.#closed = true;
      if (error instanceof HarnessServiceError) throw error;
      throw invalidContext();
    }
  }

  async terminate(): Promise<void> {
    if (this.#closed) return;
    this.#closed = true;
    this.#subprocess.kill();
    await this.#subprocess.exited.catch(() => undefined);
  }

  async #receive(): Promise<CandidateContext> {
    const output = await new Response(this.#subprocess.stdout).text();
    const exitCode = await Promise.race([
      this.#subprocess.exited,
      new Promise<number>((resolve) => setTimeout(() => {
        this.#subprocess.kill(9);
        resolve(-1);
      }, 5_000)),
    ]);
    if (exitCode !== 0) throw invalidContext();
    const payload = JSON.parse(output) as CandidateContextPayload;
    if (payload.ok !== true || payload.candidate === undefined) throw invalidContext();
    const candidate = payload.candidate;
    const profile = attributedSource(candidate.profileNarrative.name, "profile", candidate.profileNarrative.text);
    const contexts = candidate.contextSources.map((source) => attributedSource(source.name, "context", source.text));
    const anecdotes = candidate.anecdotes.map((source) => attributedSource(source.name, "anecdote", source.text));
    if (typeof candidate.resumeText !== "string" || characterCount(candidate.resumeText) > MAX_RESUME_SOURCE_CHARACTERS) throw invalidContext();
    const directFields: Record<string, string> = Object.create(null) as Record<string, string>;
    for (const [name, value] of Object.entries(candidate.directFields)) {
      if (typeof value !== "string") throw invalidContext();
      directFields[name] = value;
    }
    return Object.freeze({
      directFields: Object.freeze(directFields),
      resumeText: candidate.resumeText,
      profileNarrative: profile,
      contextSources: Object.freeze(contexts),
      anecdotes: Object.freeze(anecdotes),
    });
  }
}

if (import.meta.main && process.argv.at(-1) === "--candidate-context-worker") {
  await runContextWorker();
}
